### PR TITLE
Add code for pool affinity key

### PR DESCRIPTION
**Pull request by sinhaashish**
this PR adds the code in 
- CSI
- Rest
- GRPC
- Stor-port

### CODE REVIEW

File: control-plane/csi-driver/src/context.rs [hunk 1]
```diff
@@ -116,6 +118,35 @@ impl Parameters {
...
+    ) -> Result<Option<Vec<String>>, tonic::Status> {
+        Ok(match value {
+            Some(labels) => {
+                let mut result_vec = Vec::new();
+                for label in labels.split('\n') {
+                    if !label.is_empty() {
+                        result_vec.push(label.to_string())
+                    } else {
+                        return Err(tonic::Status::invalid_argument(format!(
+                            "Invalid label : {value:?}"
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]:
```suggestion
                            "Invalid affinity key : {value:?}"
```

File: control-plane/rest/openapi-specs/v0_api_spec.yaml [hunk 2]
```diff
@@ -2589,9 +2591,22 @@ components:
           type: object
           additionalProperties:
             type: string
+        affinitykey:
```
Comment (tiagolobocastro) [inline control-plane/rest/openapi-specs/v0_api_spec.yaml]: I wonder if we should split the topology into requested and actual topology.
Then on request for volume creation affinityKey is a vector, but on the actual volume, then it's a dict of key:value
Thoughts?  CC @sinhaashish  @Abhinandan-Purkait

File: control-plane/csi-driver/src/context.rs [hunk 3]
```diff
@@ -116,6 +118,35 @@ impl Parameters {
...
     }
 
+    /// Parses the value for affinity key
+    /// PoolAffinityTopologyKey: |
+    ///   A1
+    ///   CE
+    ///  The input value to this function is like Some("A1\nCE \n")
+    ///  and it will be parsed as {"A1": "", "CE ": ""}
+    /// and it will be parsed as ["A1", "CE "]
+
```
Comment (tiagolobocastro) [inline control-plane/csi-driver/src/context.rs]:
```suggestion
```

**Review by tiagolobocastro — APPROVED** [review]

**Review by Abhinandan-Purkait — APPROVED** [review]

File: control-plane/rest/openapi-specs/v0_api_spec.yaml [hunk 4]
```diff
@@ -2589,9 +2591,22 @@ components:
...
+             Subsequently, all other resources must match the given label "Zone: A",
+             effectively adding this requirement as an inclusion label.
+          type: array
+          items:
+            example: 'Zone'
+            type: string
       required:
         - exclusion
         - inclusion
+        - affinitykey
```
Comment (tiagolobocastro) [inline control-plane/rest/openapi-specs/v0_api_spec.yaml]:
```suggestion
```
Should not be required

File: control-plane/stor-port/src/types/v0/transport/volume.rs [hunk 5]
```diff
@@ -255,19 +255,31 @@ pub struct LabelledTopology {
     /// Inclusive labels.
     #[serde(default)]
     pub inclusion: HashMap<String, String>,
+    /// Affinity labels.
+    #[serde(default)]
+    pub affinity: HashMap<String, String>,
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/volume.rs]:
```suggestion
    pub creation_affinity_keys: Option<Vec<String>>,
```

**Review by tiagolobocastro — COMMENTED** [review]
When we're converting from internal proto to openapi, then we're loosing the affinity value anyway?
So we need to either split the request/spec types or don't use the affinity_key map and we simply move affinity_key to the inclusion on successful volume creation?

**Comment by Abhinandan-Purkait** [comment]
> When we're converting from internal proto to openapi, then we're loosing the affinity value anyway? So we need to either split the request/spec types or don't use the affinity_key map and we simply move affinity_key to the inclusion on successful volume creation?

Sounds, good to me. 👍🏻

**Comment by sinhaashish** [comment]
I have removed affinity from the `LabelledTopology`. Its just that once the volume is created the inclusion will show the key value, and value will be created at the run time.

**Comment by sinhaashish** [comment]
@tiagolobocastro @Abhinandan-Purkait PTAL , not merging it until re approved

File: control-plane/rest/openapi-specs/v0_api_spec.yaml [hunk 6]
```diff
@@ -2589,6 +2591,18 @@ components:
           type: object
           additionalProperties:
             type: string
+        affinitykey:
```
Comment (tiagolobocastro) [inline control-plane/rest/openapi-specs/v0_api_spec.yaml]: nit:
```suggestion
        affinity_key:
```

File: control-plane/stor-port/src/types/v0/transport/volume.rs [hunk 7]
```diff
@@ -255,13 +255,17 @@ pub struct LabelledTopology {
     /// Inclusive labels.
     #[serde(default)]
     pub inclusion: HashMap<String, String>,
+    /// Affinity labels.
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/volume.rs]:
```suggestion
    /// Affinity keys.
```

**Review by tiagolobocastro — APPROVED** [review]

**Comment by sinhaashish** [comment]
bors merge

**Comment by sinhaashish** [comment]
bors merge

**Comment by sinhaashish** [comment]
bors merge

**Comment by sinhaashish** [comment]
bors merge

**Pull request merged**